src/store/reducers/keymap.ts

```diff
@@ -0,0 +1,54 @@
+import { Action } from '@ngrx/store';
+
+import { Keymap } from '../../config-serializer/config-items/Keymap';
+import { KeymapActions } from '../actions';
+
+const initialState: Keymap[] = [];
+
+export default function(state = initialState, action: Action): Keymap[] {
+    switch (action.type) {
+        case KeymapActions.ADD:
+            let newKeymap: Keymap = Object.assign({}, action.payload);
+
+            newKeymap.abbreviation = generateAbbr(state, newKeymap.abbreviation);
+            newKeymap.name = generateName(state, newKeymap.name);
+
+            return [...state, newKeymap];
+        default: {
+            return state;
+        }
+    }
+}
+
+function generateAbbr(state: Keymap[], abbr: string): string {
+    let chars: string[] = '23456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('');
```

Comment:
Yes, this can be changed to const. You can read more [here](https://github.com/UltimateHackingKeyboard/agent/issues/78#issue-171215656). One is missing because, this generator generates alternative name if same name already exists, and you don't want to start with 1 in this case. Same goes for function `generateName` where `suffix ` start's with 2.